artemis-protocols/artemis-mqtt-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/mqtt/MQTTUtil.java

```diff
@@ -107,120 +155,339 @@ public static String convertMQTTAddressFilterToCoreRetain(String filter, Wildcar
       return MQTT_RETAIN_ADDRESS_PREFIX + MQTT_WILDCARD.convert(filter, wildcardConfiguration);
    }
 
-   private static ICoreMessage createServerMessage(MQTTSession session,
-                                                    SimpleString address,
-                                                    boolean retain,
-                                                    int qos) {
+   private static ICoreMessage createServerMessage(MQTTSession session, SimpleString address, MqttPublishMessage mqttPublishMessage) {
       long id = session.getServer().getStorageManager().generateID();
 
-      CoreMessage message = new CoreMessage(id, DEFAULT_SERVER_MESSAGE_BUFFER_SIZE, session.getCoreMessageObjectPools());
+      CoreMessage message = new CoreMessage(id, mqttPublishMessage.fixedHeader().remainingLength(), session.getCoreMessageObjectPools());
       message.setAddress(address);
-      message.putBooleanProperty(MQTT_MESSAGE_RETAIN_KEY, retain);
-      message.putIntProperty(MQTT_QOS_LEVEL_KEY, qos);
+      message.putBooleanProperty(MQTT_MESSAGE_RETAIN_KEY, mqttPublishMessage.fixedHeader().isRetain());
```

Comment:
Did you consider the possibility to create a MQTTMessage?

If you're not doing that now, and we plan to eventually do it.. using application properties from CORE wouldn't make it a bit more difficult to make such transition?

Reply:
I didn't seriously consider implementing a native MQTTMessage at this point. My goal in this PR was to correctly implement all the new MQTT 5 features.

We've always used message properties when converting to/from core for MQTT-specific details. This PR isn't changing that. I'm not clear on how this would make the transition to a native MQTTMessage implementation more difficult later. Can you elaborate on that?